### PR TITLE
Introduce vulkan.hpp-based version of hello_triangle.

### LINKED ISSUE #304: Vulkan HPP Framework Discussion
Hey @asuessenbach, It would be great to discuss the Vulkan HPP Framework changes before the next call :)

### Overview

> Demonstrate more use cases with the Vulkan HPP API. This may require the recreation of another framework using Vulkan HPP instead of the C style API. A selection of API and performance samples would be ported to show its benefits ~ paraphrased from previous conversations

My biggest concern with duplicating the framework is that when larger parts are refactored in one of the frameworks then these will need to be reflected in the other; the same applies for samples. Replicating existing samples would mean that contributors will need to take more factors into account - there is already an ever increasing check list on PRs.

### Is this a good direction?

I'm worried that direct ports of samples will generate more work in the future without providing much value for the developers looking at these samples. For instance, the inner workings of hello triangle have not been changed. We are simply seeing the same code written in a different style. This is similar to to-do list examples across programming languages in the sense that the added value for a developer is learning dramatically different syntax. In this case, the C style API covers all the Vulkan syntax that a developers needs and the HPP is "syntactic sugar".

It may be a better use of time to find how the HPP API can be used in comparison to the C style API when tackling specific problems in Vulkan. Are there any scenarios where we can show one API greatly simplifying concepts for developers or reducing the size of code footprints for specific features?

One example of this could be showing the benefits of: not using a shared context, current framework; using a shared context, shared pointer to a struct of common handles and using the HPP style API to not require tracking lower level handles - device, instance etc... Or possibly rethinking resource management with tracking structs of handles vs tracking hpp vulkan objects... Can hello triangle be written in a more simple way using HPP...

### Summary

Overall, I think it is great to add more HPP. It just may be more beneficial to demonstrate its pros/cons through new samples rather than maintain two copies of the current codebase

### PR DESCRIPTION
**Pull request by asuessenbach**
## Description

Add a new sample `hpp_hello_triangle`, which is a clone of the `hello_triangle` sample illustrating the usage of the C++ bindings provided by vulkan.hpp.
In order to make that work, a change is needed to the pure virtual function `Window::create_surface`, now having a `VkInstance` instead of a `vkb::Instance &` as its only argument. `AndroidWindow`, `GlfwWindow` and `HeadlessWindow` are adjusted accordingly.
As a consequence, usage of `vkb::Instance` in the `hello_triangle` sample could be removed, which reduces its use of the framework even more.

I've built and tested the new sample on Windows.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/api/hello_triangle/hello_triangle.cpp [hunk 1]
```diff
@@ -1070,9 +1068,7 @@ bool HelloTriangle::prepare(vkb::Platform &platform)
 {
 	init_instance(context, {VK_KHR_SURFACE_EXTENSION_NAME}, {});
 
-	vk_instance = std::make_unique<vkb::Instance>(context.instance);
-
-	context.surface = platform.get_window().create_surface(*vk_instance);
+	context.surface = platform.get_window().create_surface(context.instance);
```
Comment (TomAtkinsonArm) [inline samples/api/hello_triangle/hello_triangle.cpp]: Nice, makes sense

File: samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp [hunk 2]
```diff
@@ -0,0 +1,1004 @@
...
+ */
+vk::Result HPPHelloTriangle::present_image(Context &context, uint32_t index)
+{
+	vk::PresentInfoKHR present(context.per_frame[index].swapchain_release_semaphore, context.swapchain, index);
+	// Present swapchain image
+	return context.queue.presentKHR(present);
+}
+
+/**
+ * @brief Initializes the Vulkan frambuffers.
```
Comment (TomAtkinsonArm) [inline samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp]: typo: framebuffers

Comment (TomAtkinsonArm) [reply samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp]: Same on line 947 `hello_triangle.cpp`

**Review by TomAtkinsonArm — COMMENTED** [review]
Port from `hello_triangle` LGTM

Experiencing a crash on S21 when retrieving VkSurface. Not sure why this is caused yet

**Comment by TomAtkinsonArm** [comment]
> Experiencing a crash on S21 when retrieving VkSurface. Not sure why this is caused yet

Works now, error was due to own fault :)

**Review by gary-sweet — CHANGES_REQUESTED** [review]
I'm seeing build errors for d2d platforms:

In file included from framework/platform/unix/unix_d2d_platform.cpp:28:
framework/platform/unix/direct_window.h:39:23: error: 'virtual VkSurfaceKHR_T* vkb::DirectWindow::create_surface(vkb::Instance&)' marked 'override', but does not override
  virtual VkSurfaceKHR create_surface(Instance &instance) override;
                       ^~~~~~~~~~~~~~
In file included from framework/platform/unix/direct_window.cpp:18:
framework/platform/unix/direct_window.h:39:23: error: 'virtual VkSurfaceKHR_T* vkb::DirectWindow::create_surface(vkb::Instance&)' marked 'override', but does not override
  virtual VkSurfaceKHR create_surface(Instance &instance) override;

File: samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp (hunk 2 above)
Comment (asuessenbach) [reply samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp]: Resolved with https://github.com/KhronosGroup/Vulkan-Samples/pull/294/commits/1e497ba1a21aaf4e7c9b63ccbe9e7ce93915830d

**Comment by asuessenbach** [comment]
The platform/unix override of `create_surface` is the only one that really uses the `vkb::Instance` argument, and not just the `VkInstance` from it (via `get_handle()`): the function `get_suitable_gpu` is called here.
One could either
1. replicate that functionality locally, or
2. extent the signature of `create_surface` to get a `VkPhysicalDevice` as a second argument, or
3. revert the changes on `create_surface` in this PR, and add some HPP-version of `vkb::Instance`,`vbk::PhysicalDevice`, as well as some HPP-version of every flavor of `vkb::Window` instead

While approach 1. would be very local, some code would be replicated and thus would exist redundantly (even though, it's very little code). One could maybe introduce `get_suitable_gpu` as a helper function, working on a `VkInstance`, instead.

Approach 2. would be pretty small, but would require some work on the other platforms where it is not needed. Moreover, the hello_triangle sample would either need to keep the usage of vkb::Instance or use its own version of `get_suitable_gpu`.

Finally, approach 3. would need some substantial work, replicating lots of code. But depending on the long-term goals of adding hpp-based samples, that work might be needed nevertheless.

So, any advice how to approach this?

File: samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp [hunk 3]
```diff
@@ -0,0 +1,1004 @@
...
+		throw std::runtime_error("Required instance extensions are missing.");
+	}
+
+	std::vector<vk::LayerProperties> supported_validation_layers = vk::enumerateInstanceLayerProperties();
+
+	std::vector<const char *> requested_validation_layers(required_validation_layers);
+
+#ifdef VKB_VALIDATION_LAYERS
+	// Determine the optimal validation layers to enable that are necessary for useful debugging
+	std::vector<const char *> optimal_validation_layers = vkb::get_optimal_validation_layers(supported_validation_layers);
```
Comment (SaschaWillems) [inline samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp]: This doesn't compile for me when enabling validation layers. The `get_optimal_validation_layers` function takes a vector if `VkLayerProperties`, which is not compatible with the hpp's `vk::LayerProperties` array. 

I guess we need an additional overload for this to work.

**Review by SaschaWillems — CHANGES_REQUESTED** [review]
Thanks for contributing this. Having up-to-date samples for Vulkan.hpp is a very welcome addition 👍 

Except for the compilation failure when enabling the validation layers, this looks good to me on Windows, but as with D2D, this fails for on Android.

**Comment by SaschaWillems** [comment]
> I'm seeing build errors for d2d platforms:
> 
> In file included from framework/platform/unix/unix_d2d_platform.cpp:28:
> framework/platform/unix/direct_window.h:39:23: error: 'virtual VkSurfaceKHR_T* vkb::DirectWindow::create_surface(vkb::Instance&)' marked 'override', but does not override
> virtual VkSurfaceKHR create_surface(Instance &instance) override;
> ^~~~~~~~~~~~~~
> In file included from framework/platform/unix/direct_window.cpp:18:
> framework/platform/unix/direct_window.h:39:23: error: 'virtual VkSurfaceKHR_T* vkb::DirectWindow::create_surface(vkb::Instance&)' marked 'override', but does not override
> virtual VkSurfaceKHR create_surface(Instance &instance) override;

@gary-sweet : Maybe we can add a D2D target to the CI pipeline? This way we could easily catch build errors that only affect D2D builds.

File: samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp (hunk 3 above)
Comment (asuessenbach) [reply samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp]: Resolved with https://github.com/KhronosGroup/Vulkan-Samples/pull/294/commits/3e62a9f5a339e539e4833c0594e0a8376991286d.

**Comment by TomAtkinsonArm** [comment]
> Maybe we can add a D2D target to the CI pipeline? This way we could easily catch build errors that only affect D2D builds.

I can attempt this, I think last time I tried there where some dependencies that where not available for it to compile

**Comment by gary-sweet** [comment]
> > Maybe we can add a D2D target to the CI pipeline? This way we could easily catch build errors that only affect D2D builds.
> 
> I can attempt this, I think last time I tried there where some dependencies that where not available for it to compile

I thought we already had this, but maybe it was only in the old gitlab version.

**Comment by gary-sweet** [comment]
I'm now seeing:

framework/platform/unix/direct_window.cpp: In member function 'virtual vk::SurfaceKHR vkb::DirectWindow::create_surface(vk::Instance, vk::PhysicalDevice)':
framework/platform/unix/direct_window.cpp:373:40: error: 'mode_props' was not declared in this scope
  surface_create_info.displayMode     = mode_props.displayMode;
                                        ^~~~~~~~~~
framework/platform/unix/direct_window.cpp:373:40: note: suggested alternative: 'mode_t'
  surface_create_info.displayMode     = mode_props.displayMode;

**Comment by gary-sweet** [comment]
I think you should be able to do a D2D build by forcing VKB_WSI_SELECTION STREQUAL to "D2D" in CMake (on Linux at least).

**Comment by asuessenbach** [comment]
@gary-sweet Should be fixed now.

**Comment by gary-sweet** [comment]
> @gary-sweet Should be fixed now.

Indeed it is - thanks.

I am seeing one slight oddity now that I can run it. When I terminate the app (using the ESC key for example), I see:
```
terminate called without an active exception
Aborted
```

I've run a bunch of the other samples and none of them exhibit this behaviour on shutdown.

**Comment by asuessenbach** [comment]
@gary-sweet Strange thing!
Can you spot where that terminate comes from? Maybe some part of the DirectWindow that would need a transcoded version as well?

**Comment by gary-sweet** [comment]
> @gary-sweet Strange thing!
> Can you spot where that terminate comes from? Maybe some part of the DirectWindow that would need a transcoded version as well?

All of the work in main() has completed before the message is printed. It appears when main() returns. The debugger isn't much help either, it spots the exception but the stack-trace says "[Unknown/Just-In-Time compiled code]"

The odd thing is that this is the only sample that has this behaviour - all the others seem ok, and they all use D2D on my platform.

**Comment by asuessenbach** [comment]
@gary-sweet  Could you please give it another try?
I somehow missed to destroy the instance... and nobody complained.

**Comment by gary-sweet** [comment]
> @gary-sweet Could you please give it another try?
> I somehow missed to destroy the instance... and nobody complained.

Yep. That fixes it. Thanks

**Review by gary-sweet — APPROVED** [review]

**Review by SaschaWillems — APPROVED** [review]
LGTM 👍

**Review by gpx1000 — APPROVED** [review]
LGTM;

**Review by TomAtkinsonArm — APPROVED** [review]

**Comment by marty-johnson59** [comment]
Merging now that we have 4 approvals.  LMK if we need to revert.

mcj

**Pull request merged**